cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/AbstractLogPublisher.java

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.logging.appender;
+
+import co.cask.cdap.common.service.AbstractRetryableScheduledService;
+import co.cask.cdap.common.service.RetryStrategy;
+import kafka.producer.KeyedMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+
+/**
+ * A class that continuously serializes logs from a queue and publishes them.
+ *
+ * @param <MESSAGE> the type of message used in the in-memory buffer, before publishing
+ */
+public abstract class AbstractLogPublisher<MESSAGE> extends AbstractRetryableScheduledService {
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractLogPublisher.class);
+
+  private final int queueSize;
+  private final BlockingQueue<LogMessage> messageQueue;
+  private final List<MESSAGE> buffer;
+  private volatile boolean failed;
+  private volatile Thread publishThread;
+
+  public AbstractLogPublisher(int queueSize, RetryStrategy retryStrategy) {
+    super(retryStrategy);
+    this.queueSize = queueSize;
+    this.messageQueue = new ArrayBlockingQueue<>(queueSize);
+    this.buffer = new ArrayList<>(queueSize);
+  }
+
+  /**
+   * Creates a {@link MESSAGE}, which the type is dependent on the implementation. This is buffered in-memory before
+   * being published.
+   *
+   * @param logMessage the log message to prepare for publishing
+   * @return a {@link MESSAGE}, prepared for publishing.
+   */
+  protected abstract MESSAGE createMessage(LogMessage logMessage);
+
+  /**
+   * Responsible for publishing a list of log messages.
+   *
+   * @param logMessages the list of log messages to be published.
+   */
+  protected abstract void publish(List<MESSAGE> logMessages) throws Exception;
+
+  /**
+   * Enqueues a log message for publishing.
+   *
+   * @param logMessage the log message to enqueue for publishing
+   * @throws InterruptedException if interrupted while waiting to enqueue (waiting is done in case the queue is full)
+   */
+  public final void enqueueForPublishing(LogMessage logMessage) throws InterruptedException {
```

Comment:
Just all it `addMessage`. Calling it enqueue kind of exposing the implementation.